src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -442,30 +451,30 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
                 return null;
             }
 
-            if (csharpResponse.Edits == null)
+            if (csharpResponse.Edits is null)
             {
                 Assumes.NotNull(csharpResponse.Tokens);
 
-                // C#'s edit handler returned to us a full token set, so we don't need to do any additional work.
-                // This should ideally happen rarely as it indicates C# had trouble with cache retrieval.
-                return new SemanticTokens { ResultId = csharpResponse.ResultId, Data = csharpResponse.Tokens.ToImmutableArray() };
+                // C#'s edits handler returned to us a full token set, so we don't need to do any additional work.
+                // This may indicate that C# had trouble with cache retrieval or that the previous tokens were
+                // 0-length, since C# doesn't cache 0-length token sets.
+                return new SemanticTokensResponse(csharpResponse.ResultId, csharpResponse.Tokens.ToArray(), csharpResponse.IsPartial);
             }
 
             Assumes.NotNull(csharpResponse.Edits);
 
             if (!csharpResponse.Edits.Any())
             {
                 // If there aren't any edits, return the previous tokens with updated resultId.
-                return new SemanticTokens { Data = previousCSharpTokens.ToImmutableArray(), ResultId = csharpResponse.ResultId };
+                return new SemanticTokensResponse(csharpResponse.ResultId, previousCSharpTokens.ToArray(), csharpResponse.IsPartial);
             }
 
             var updatedTokens = ApplyEditsToPreviousCSharpDoc(previousCSharpTokens, csharpResponse.Edits);
-
-            return new SemanticTokens { Data = updatedTokens, ResultId = csharpResponse.ResultId };
+            return new SemanticTokensResponse(csharpResponse.ResultId, updatedTokens.ToArray(), csharpResponse.IsPartial);
         }
 
         // Internal for testing
-        internal static ImmutableArray<int> ApplyEditsToPreviousCSharpDoc(
+        internal static int[] ApplyEditsToPreviousCSharpDoc(
```

Comment:
What drove this change?

Reply:
Two reasons:
1) We now pass around `SemanticTokenResponse` instead of `SemanticTokens` (due to the the need to include `IsFinalized`) and the former takes `int[]` while the latter takes `ImmutableArray<int>`.
2) I believe ImmutableArray is a tiny bit more expensive than regular arrays since it's a wrapper class? So I thought it might result in very minor perf gains here.